### ISSUE TITLE
python modules in the repository

### ISSUE THREAD

**Issue by pavlis** [wangyinz/mspass#26]
I've started writing an obspy2mspass python function that will take an obspy Trace object and construct a TimeSeries object from it.   The issue I wanted to ask is two-fold:

1. Should we start a different "module" for this kind of thing or integrate the code with mspasspy?  I'm not sure if the later is actually possible as it would mix binary .so (dylib) with python scripts.   Might be possible with cython, but that might get even uglier.   
2.  How should the repository be laid out to handle a library of python procedures?

**Comment by wangyinz** [comment]
That is the other thing I was thinking about today. I was trying to find a good example on how to make CMake work with any build systems of Python. Not sure that there is anyone actually doing such a thing. That's why in https://github.com/wangyinz/mspass/issues/17#issuecomment-562760031 I prefer not to have another python script to handle the template. I think for now, we should always put C++ code in `/cxx` and Python code in `/python`.

I am still in the phase of learning the best practice of organizing a python project, so I don't have a good answer to that yet. I think for now, it is OK to just put it under `/python`. I will reorganize it once I come up with a good plan. I guess it will be fine at this early stage of development.

**Comment by pavlis** [comment]
Keep in mind I'm at about the 200 level in python programming - I know most of the features of the language but am no totally fluent.   That is a caveat that this may be off base.

Was reading about "modules" and "packages" in [this online source](https://realpython.com/python-modules-packages/).   Hadn't quite understood the distinction between the two before.   Made me realize we might want to organize some of this stuff into a package with mspass as the top level name.  That source says this is easy for pure python code; you just put the code for each module in subdirectories.   obpsy does this, which is why they have obspy.core, obspy.db, etc.   Seems to me we should do that too to keep the namespace cleaner.  I'm a bit unsure how to do that with the C code, but doubt it is that hard.   Just need to know he tricks

Here is suggested package.module starting organization:
mspass - let this become the package name.   Alternatively, we could call it mspasspy to make it clear that is the python package.  mspasspy is probably better.

**mspasspy.Cseis** or just **mspasspy.seismic** or maybe **mspasspy.core** ala obspy
This would contain all seismic related code produced from wrapped C++.   One model is the current mspasspy module.  We might consider splitting that module into a seismic and utility (or some similar name) to follow the C source tree organization.  There is logic in that as the utility components have use beyond just seismology.
**mspasspy.parallel** 
Python scripts suitable for map-reduce.  The example.py script you have is a starting point.
**mspasspy.processing**
Might want a separate module for processing algorithms not linked to spark.   I'm thinking these would be normal function or classes that could process one object at a time.  The idea is map-reduce scripts would typically have calls to these functions/class-methods.

The other key advantage this would provide is a cleaner method for add ons by others as the system develops.  We could use the antelope model and have a **mspasspy.contrib** or create other generic names for common elements.  e.g. I could see evolution to **mspasspy.graphics**, **mspasspy.simulation**, or any number of others.

**Comment by wangyinz** [comment]
This needs to be the organization for the python directory. I prefer to still call it mspasspy, since mspass should refer to the whole thing including MongoDB and Spark. The C++ code need to have a different name, since it may not be seismic code only. Maybe we should call it **mspasspy.cmodule** or just **mspasspy.cm**.  Then, we can just follow the C source tree organization, and have submodules **mspasspy.cm.seismic** and **mspasspy.cm.utility**. Not sure what we wanna do with the **mspasspy.cm.algorithm** module, since that overlaps with  **mspasspy.processing** in functionality. Maybe we should name the same which will be **mspasspy.cm.processing** and **mspasspy.processing**.

**Comment by wangyinz** [comment]
I have uploaded current version of the mspasspy package to PyPI as a test. There is a couple issue that worth noting here. I ran following commands to create the distribution and upload it:
```
python3 setup.py sdist bdist_wheel
mv dist/mspasspy-0.0.1-cp36-cp36m-linux_x86_64.whl  dist/mspasspy-0.0.1-cp36-cp36m-manylinux2014_x86_64.whl
python3 -m twine upload dist/*
```
Note that if doing this within the Docker container, we will need to install the following packages:
```
pip3 install --upgrade twine keyrings.alt
```
The issue is that I have to change the name of the generated distribution due to PyPI's support of compiled package (refer to [PEP 599](https://www.python.org/dev/peps/pep-0599/)). So, what I have above it totally a hack. Preferably, we will want to setup a Travis workflow to do the distribution with the Docker image provided [here](https://github.com/pypa/manylinux). There are more setups needed though as we have a lot of dependencies to be resolved.

**Comment by pavlis** [comment]
I now have a specific problem that requires a solution to this issue.   We have the base mspasspy module produced from the C++ code and wrappers. 

We very much need to start the process of construction of basic functions to load static data (what antelope calls dbmaster) and waveforms.   In my opinion these need to be split into a different module that splits these functions into a functional block.   This particular one should probably be called mspass.import.

How do we build this structure?   A complication is these will likely require a mix of C++ and python code.

**Comment by wangyinz** [comment]
Oh, I thought we are only developing the reader and writer in Python. Since we have the C++ wrapped in Python already, can't we just do the import with the MongoDB Python API? What is the need to C++ code here? We could build a similar import module in our C++ API using MongoDB C++ API, but that should be a separate development, right?

**Comment by wangyinz** [comment]
If we are following the directory structure that we decided earlier, the **mspasspy.import** module should be placed in `/python/import/`.

**Comment by pavlis** [comment]
Ok, that is a clean structure for a python module and agrees with the web source at the head of this issue discussion.  

 I think I confused the issue with the sentence about mixing C++ and python, but it really comes to a head here.   The issue is that the easiest path to solution for me to write the parser for the yaml representation of the Metadata is to write a C++ program using that package you added earlier (don't remember the name immediately, but the C++ code that parses yaml).  This has a scalability issue I haven't quite figure out how to deal with yet.   The way I wrote the antelope intermediate file is that the yaml file has one block in a list for each seismogram in a data file.   The example I just ran had 1913 seismograms from the first year of USArray data limited to segments defined by P picks made at ANF.   That is a decent starting point, I think.   However, later the number of seismogram could be millions if I put everything in year files and didn't restrict the conversion to the largest events.   100000, at least, would be routine.   I can eat the header data, but not all the waveform data at once.   

The problem this raises is that writing the yaml parser should be easy.  I think I can do it initially the lazy way and just have the parser build an std::vector<mspass::Metadata> container or probably better to use a list as it wouldn't require the data to be in a single memory block.   If you look in the cxx/python directory you will see a function I wrote in C++ that can be called from Python to convert Metadata to a python dict that can be used for CRUD operations in Mongo.  This converter crosses the C++ to python boundary at a nasty point.   Reading the yaml and binary files the antelope converter produced is easiest to do in C++, but the database transactions are best done in python.   

I think I only have two options:

1. parse the yaml file in python - presumably there are good libraries out there - and create a python converter to read the sample data and save the result in MongoDB.   
2.  crack the C++ database interface and have this converter be pure C++ with the only objective being to load antelope data. 

I suspect #1 is the easier path to solution.  Any suggestions?

**Comment by pavlis** [comment]
[PyYAML](https://pyyaml.org/wiki/PyYAML) seems the likely library for a pure python solution.   Any experience with this?

**Comment by wangyinz** [comment]
Yes, I can confirm that is the common choice. Should be able to just do a `pip install pyyaml` and then use it as `import yaml`.

**Comment by wangyinz** [comment]
I agree it is easier to do it in Python with its own yaml parser. This will make it easier to run in parallel with Spark.

**Comment by pavlis** [comment]
Sorry for the rambling comment above.  The first step of this turned out to be absurdly simple.   Once I installed pyyaml as you noted above I copied the yaml file created from the 2004 USArray data yesterday.   Reading was trivial:
file=open('2004test.yaml')
d=yaml.load(file,Loader=yaml.FullLoader)

Could get all the metadata form any object with a simple index.  e.g. to get #5 
print(d[5])
That yielded for this example:
{'Object_number': 5, 'iphase': 'P', 'phase': 'pP', 'sta': 'DEC', 'evid': 171954, 'foff': 3360120, 'ns': 28001, 'nsamp': 28001, 'orid': 279116, 'delta': 0.05, 'mb': 5.8, 'ms': 6.2, 'receiver_elev': 0.519, 'receiver_lat': 34.2535, 'receiver_lon': -118.3338, 'source_depth': 10, 'source_lat': -19.374, 'source_lon': 169.631, 'starttime': 1082059883.4841}

You can get a specific entry by a second key:
print(d[5]['delta'])
yields
0.05

Trying to figure out how to make this more scalable by reading one object's data at a time.   That is I followed the simple online example that eats up the whole file.   I think the package has ways to read one block at a time.   The docstring says so anyway, but I may need to change the format a bit to wrap each data's metadata as a single document (trying to figure out what a "document" means in the yaml format).  

Anyway, this will provide an excellent way to load a large volume of test data for benchmarks.   I have 2004-2006 USArray data already loaded in a form that should be easy to convert and load into MongoDB by this process.

**Comment by wangyinz** [comment]
I think the document in yaml refers to sections delineated by `---`. You can find some examples [here](https://yaml.org/spec/1.1/#id857577).

**Comment by pavlis** [comment]
I just made an executive decision on how to initially structure the python code I've developed so far while thinking about how these should be readily extended.  

Top of the chain, as I think you were suggesting in this chain is mspass/python
Under that is mspass/python/mspass
Below that are presently these directories:  importer and mongodb.   

The logic of this is the current chain has these three modules loaded as follows in verbose form:
import mspass.importer.seispp
import mspass.importer.obspy
import mspass.mongodb.seispp

Presumably we would normally recommend this kind of thing:
import mspss.importer.obspy as iop 
or whatever symbol one wanted.   

I just checked all that in.  Please feel free to rearrange if you don't like that structure.   You can and should control the organization of the repository.

Related was that I could not make any of that work without setting the PYTHONPATH environment variable to mspass/python.   The setup.py script does not seem to do that configuration unless I missed a key incantation somewhere.

**Comment by pavlis** [comment]
The PYTHONPATH approach is messing something up badly.   The behavior I'm seeing is mysterious, because sometime it has worked but mostly (alway this morning) I get this message after setting PYTHONPATH to src/mspass/python
pavlis-home-linux:~/testing/seisppimport> setenv PYTHONHOME /home/pavlis/src/mspass/python
pavlis-home-linux:~/testing/seisppimport> python
Fatal Python error: Py_Initialize: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00007f8bfb54a740 (most recent call first):
Abort (core dumped)
pavlis-home-linux:~/testing/seisppimport> setenv PYTHONHOME /home/pavlis/src/mspass/python/mspss
pavlis-home-linux:~/testing/seisppimport> python
Fatal Python error: Py_Initialize: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00007f8f6cb62740 (most recent call first):
Abort (core dumped)

Noting that before I ran the setenv PYTHONPATH was not defined at all.  

google searches suggests this may be related to some hidden path that is used when PYTHONPATH is undefined and somehow get lost when set as I did.   Haven't tried this on the Mac - the above was with my ubuntu machine.   

A working hypothesis for a fix is something needs to be configured in setup.py, but that may be dead wrong.

**Comment by wangyinz** [comment]
I finally got the `setup.py` working at an acceptable degree to me. Basically, all Python components will be built and installed with 
```
pip3 install ${root_dir_of_mspass}
```
Everything is under the `mspasspy` name space now, and I put the cxx library as a sub-module named `ccore`. I didn't change the name of the two other modules that you've been working on (`mspasspy.importer` and `mspasspy.mongodb`), but I do expect to change them. Given that these modules are really just a number of tools to transfer\convert data, I think we should just put them all under a `mspasspy.util` sub-module.  

Another issue that I found is that a import call to the current Python modules will kinda contaminate the namespace. For example, by calling `import mspasspy.importer.seispp`, objects like `mspasspy.importer.seispp.MongoClient` become available to current name space, which is confusing. I guess there should be similar rules on how the import lines should be used in a module, just like the using namespace line in C++.

**Comment by pavlis** [comment]
You are in charge of the organization.   You should see when this new stuff gets checked in.  Hoping that will be today.   Ran across a couple additional issues I had to fix in the C++ api  - typical of course when you in a new development and why I don't think it is ever possible to build perfect API to anything without using it a lot first. 
For the namespace contamination please educate me.  Realize I'm not (yet) a fully functional python coder.

**Comment by pavlis** [comment]
The new setup.py didn't work for me, but I think it is totally related to one thing: In your previous comment you said had this entry:
``
pip3 install ${root_dir_of_mspass}
``
Do I literally need a root_dir_of_mspass defined or is that the install directory of cmake?   
This is in the realm of something I've never cracked about python, which is where packages are actually installed.   Right now on this ubuntu machine it is clear they are going into the default  of /usr/local/python3 something.   Seems to me it would be better to isolate the mspass stuff in the install directory defined in the cmake configuration.   In fact, that would be essential if I were doing this development on one of the hpc machines because I could write to /usr/local.   

Please educate me on this.   I'm unable to test any of my new stuff until I sort this out.

**Comment by pavlis** [comment]
A completely different issue came up when i ran a git pull on the repository with your changes yesterday.  Scared the hell of of me for a minute as the auto merge made the seispp.py file I was working on vanish.  My panic didn't last long because I found mspasspy parallel to the old mspass directory and there was the file I'd been working on.   git did it right, but that was a bit unexpected.

**Comment by wangyinz** [comment]
After digging into Python documents, I realized that the namespace thing is not a problem at all. You are using it correctly. I think that is why `from foo import bar` is the recommended way of using any modules. 

For the pip install command, the what I mean by `${root_dir_of_mspass}` is actually the root of this repository (where `setup.py` is located). You just need to give the path as an argument (like how I did in the Dockerfile). If you don't want to install it as a system-wide package, you can add a `--user` flag to the `pip3 install` command, which will install everything into a `.local` directory under the home directory of current user. 

You are correct that both of these are kind of inconvenient for development. I think the easiest way to test the code is to not install it at all. You can just go to the directory where you can see the `mspasspy` folder (i.e. `cd ${root_of_this_repo}/python`) , and run Python interpreter there. Because python will automatically add current path to its path, it will be able to see the `mspasspy` namespace. The only thing is that you will need to manually copy the `ccore` sub-module compiled by CMake to `${root_of_this_repo}/python/mspasspy/` to make the `mspasspy.ccore` module available in this arrangement.

Yeah, sorry about that. I have to change the name of the directory as that is how Python defines its namespace. I thought that you are aware of that when I said that I changed the name of the modules... Guess that is too implicit.

**Comment by pavlis** [comment]
That workaround works fine for now.   Now I can try to debug this stuff.

**Comment by pavlis** [comment]
As the conversations above show I've been working on python mongo core read and write utilities a lot in the past week.  I just pushed a working version (well it was for me anyway) to the master branch.   All that should have been on a different branch, but I had to deal with some messy tangling of changes and found it easier to just merge into the master branch.  

I would like you to have a look at what I've done here, Ian.   I made some major design decision on how to make these things you may question.   We may need to have a conversation on the phone about all this early this week.   There is a lot here to absorb, BUT what I'm most proud of with this design are two things:

1. The user api is really simple.   The module has a bunch of functions, but the main ones a user would see are dbload3C and dbsave3C.   save has options for what is saved to allow updates of only metadata or only the sample data.  Soon we'll need a dbload and dbsave for TimeSeries data.   Probably will want a dbload_obspy and dbsave_obspy too, which would probably be easier with converters back and forth from TimeSeries.
1. Haven't tested every block, but my aim is to make dbload3C and dbsave3C never throw an an exception that isn't handled internally.   The handlers for both functions all log error messages to the ErrorLogger object in Seismogram AND before returning post any errors found in the log to an elog collection (I think is the name) stored in MongoDB.   This should make for bombproof processing in Spark.   For this reason, by the way, I think we will want to always convert obspy data to TimeSeries objects before using them in mspass.   obspy will not have the functionality as it is designed to be interactive, which this is not.   That might be an important thing to stress in any talk about mspass to contrast it as complementary to obspy - the hpc, batch processing obspy. 

There is a test program in the python/test directory called testreadwrite.py.   It worked for me when the ccore library was copied to mspass/python.   Hopefully, we can eventually get this module install thing sorted out to make that cleaner.   Maybe it will be already with the docker container.

Oh, I need to get you the raw data from 2004 and 2005 USArray data that I imported for these tests.  In fact, the testreadwrite.py script assumes those are loaded and loaded in a particular way in MongoDB.  If you want to try the test live you (Ian) can find them on the IU HPC system on dcwan.  This is the path:  /N/dcwan/projects/pavlis/USArrayParticleMotions/seispp_files.   Copy the files 2004export.* and 2005export.*   The yaml files are a yaml representation of the Metadata and the d3C files are raw fwrite dumps of the sample data in real64/double IEEE format.  Use the function in mspasspy/importers/seispp.py to load these data files into MongoDB.   It will be a good test of the docstrings to see if you can figure out how to use them.

**Comment by pavlis** [comment]
A different topic to add to the above.  There are A LOT of obvious core utilities that need to be added to the io library or perhaps another module under mongodb.   One of the things i quickly learned in running that test program is how easy it is to accumulate garbage in a MongoDB collection.  As we gain experience with this we will need to develop some housecleaning tools to help users manage the wf collection and make a useful dump of the error log collection.   

A design decision we need to make is how we will separate versions of an object that have undergone different steps in processing?  That is such a different and higher level topic, I am going to create a new issue on the topic.

**Comment by wangyinz** [comment]
Wow, that is a lot of progress. I will independently test out the loader and saver. I am way overdue to catch up on your work, but it is good to pick it up right now as I need to prepare the material for the upcoming meeting.

Yes, I realized the clean-up function will be important after messing with the database a little bit. I think it might just be an issue with developers. For users, it might be enough to just have two separate collections (or even databases) like wf and wf_tmp, where the former is for production and latter is a scratch space for experiments. Then, we just need to figure out a way to clean up the production ones according to error logs.

**Comment by pavlis** [comment]
I could use some suggestions on how to organize a set of stuff I've been building for what I think could be characterized as preprocessing.   Some functionalities the completed module could and should have are:
1.  Load station xml data downloaded by web services to build site and channel collections (done and solid)
2.  Index miniseed data.  This is a necessary preprocessing step to take raw data assembled from an array of sources before doing what would best be called an import into mspass.   There are separate threads on this topic here #74 
3.  Various forms of housekeeping functions to map raw data to a form usable by mspass will be needed.   Much of the current focus is a particular example, but there are likely other related tools that will need to developed that fit in this category. The current problem to build our full up usarray test data set is the obnoxious problem of mapping seed net:sta channel codes to data tagged with only a sta field in antelope processing.   The particular example of the day is css3.0 "arrival" table data that need to be linked to waveform data tagged with net:set.   If we don't do that right we will link the wrong waveforms to the wrong arrival times.   Details aren't the point for this  issues page. The point is there will be a growing class of (often custom) tools for cleaning up metadata problems.   What should we call these and where should they go in the python tree?
4.  I'm ultimately aimed at writing an import function, BUT it probably should be fully integrated into mspass and (a) implement history and (b) be runable as a map function in spark.  It may thus need a different home in the python tree. 
5.  There are a set of functions we need to develop for managing source metadata.   We have a few things already that are currently part of the Database class.  I'm not sure that class is correctly designed.
6.  Others?

I think a key point is we need to think about this now before the code base gets too ponderous to change things easily.   I think two key questions are:
1.  What functionality should be part of the Database class and what functionality should be put into a suite of functions?
2.  What structure should the modules for this type of thing have?  Right now we don't have an clear design for python module organization.   As you noted last week ccore may need some fragmentation too.  Can we mix pybind11 generated modules with python modules in the same directory?   I think all that is needed is to give the C++ generated library and the python code files different names right?  (e.g. if we had a dbcpp module under a db directory and had a "preprocessing.py" module in the same directory one could use constructs (I think) like:  import mspasspy.db.dbcpp and import mspasspy.db.preprocessing right?)

Anyway, let's have a discussion here as appropriate for the record in combination with offline verbal discussions in our weekly calls.

**Comment by wangyinz** [comment]
> 3. Various forms of housekeeping functions to map raw data to a form usable by mspass will be needed. What should we call these and where should they go in the python tree?
> 4. I'm ultimately aimed at writing an import function, BUT it probably should be fully integrated into mspass and (a) implement history and (b) be runable as a map function in spark.  It may thus need a different home in the python tree.
> 5. There are a set of functions we need to develop for managing source metadata.   We have a few things already that are currently part of the Database class.  I'm not sure that class is correctly designed.
> 1. What functionality should be part of the Database class and what functionality should be put into a suite of functions?

These questions are pretty much asking how we should arrange the database module. I think we should move the discussion of API design to our google doc (I already created). We definitely should turn it into offline discussions.

> 2. What structure should the modules for this type of thing have?  Right now we don't have an clear design for python module organization.   As you noted last week ccore may need some fragmentation too.  Can we mix pybind11 generated modules with python modules in the same directory?   I think all that is needed is to give the C++ generated library and the python code files different names right?  (e.g. if we had a dbcpp module under a db directory and had a "preprocessing.py" module in the same directory one could use constructs (I think) like:  import mspasspy.db.dbcpp and import mspasspy.db.preprocessing right?)

I think that will work, but we probably want to limit the mix of C++ and Python modules. It will be confusing to the users since the C++ bindings will have some subtle difference that will be unexpected to the users.

**Comment by wangyinz** [comment]
Just a note here. After re-structure the pybind11 code, it is pretty straightforward to mix Python and C++ code within a module - just put them under the same directory.

**Comment by wangyinz** [comment]
Checked by Codex

This looks like a good manual-close candidate. The repository now has the Python package layout established under `python/mspasspy/`, including Python modules alongside the bound `ccore` package. That appears to answer the original repository/module-layout question.

**Comment by wangyinz** [comment]
Closing this as completed/superseded by the established package layout.

The Python code is now organized under `python/mspasspy/`, with the bound `ccore` package and the Python `db`, `io`, `algorithms`, `preprocessing`, and `util` modules living in a single installable namespace. This resolves the repository/module-organization question raised in the issue.

Specific future package moves should be proposed separately against the current tree.
